kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java

```diff
@@ -125,18 +133,36 @@ public void appendRows(TableName tableName, List<Object[]> rows, String streamNa
                 logger.trace("Sending records to Storage API writer...");
                 ApiFuture<AppendRowsResponse> response = writer.append(jsonArr);
                 AppendRowsResponse writeResult = response.get();
-                logger.trace("Received response from Storage API writer...");
 
+                logger.trace("Received response from Storage API writer...");
                 if (writeResult.hasUpdatedSchema()) {
                     logger.warn("Sent records schema does not match with table schema, will attempt to update schema");
                     //TODO: Update schema attempt Once
                 } else if (writeResult.hasError()) {
                     Status errorStatus = writeResult.getError();
                     String errorMessage = String.format("Failed to write rows on table %s due to %s", tableName, errorStatus.getMessage());
-                    if (BigQueryStorageWriteApiErrorResponses.isMalformedErrorCode(errorStatus.getCode())) {
-                        // Fail on data error
-                        throw new BigQueryStorageWriteApiConnectException(tableName.getTable(), writeResult.getRowErrorsList());
-                        //TODO: DLQ handling
+                    if (BigQueryStorageWriteApiErrorResponses.isMalformedRequest(errorMessage)) {
+                        List<RowError> rowErrors = writeResult.getRowErrorsList();
+                        mostRecentException = new BigQueryStorageWriteApiConnectException(tableName.getTable(), rowErrors);
+                        if (getErrantRecordHandler().getErrantRecordReporter() != null) {
+                            //Routes to DLQ
+                            List<Object[]> filteredRecords = sendBadRecordsToDlqAndFilterGood(rows, convertToMap(rowErrors), mostRecentException);
+                            if (filteredRecords.isEmpty()) {
+                                logger.info("All records have been sent to Dlq.");
+                                return;
+                            } else {
+                                rows = filteredRecords;
+                                logger.debug("Sending {} filtered records to bigquery again", rows.size());
+                                jsonArr = new JSONArray();
+                                for (Object[] item : rows) {
+                                    jsonArr.put(item[1]);
+                                }
+                            }
+                        } else {
+                            // Fail if no DLQ
+                            logger.warn("DLQ is not configured!");
+                            throw new BigQueryStorageWriteApiConnectException(tableName.getTable(), rowErrors);
```

Comment:
This code segment seems to be repeated. Can we move this out to a common method and call that here to avoid duplicacy?